copier/copier.go

```diff
@@ -0,0 +1,1367 @@
+					if err != nil {
+						return errors.Errorf("error resolving hardlink target path %q under root %q", hdr.Linkname, request.Root)
+					}
+				} else {
+					linkTarget, err = resolvePath(targetDirectory, filepath.Join(targetDirectory, filepath.Dir(hdr.Name), hdr.Linkname), nil)
+					if err != nil {
+						return errors.Errorf("error resolving hardlink target path %q under root %q in directory %q", hdr.Linkname, request.Root, filepath.Dir(hdr.Name))
+					}
+				}
+				err = os.Link(linkTarget, path)
+				if err != nil && os.IsExist(err) {
+					if err = os.Remove(path); err == nil {
+						err = os.Link(linkTarget, path)
+					}
+				}
+			case tar.TypeSymlink:
+				err = os.Symlink(hdr.Linkname, path)
+				if err != nil && os.IsExist(err) {
+					if err = os.Remove(path); err == nil {
+						err = os.Symlink(hdr.Linkname, path)
+					}
+				}
+			case tar.TypeChar:
+				err = mknod(path, chrMode(mode), int(mkdev(devMajor, devMinor)))
+				if err != nil && os.IsExist(err) {
+					if err = os.Remove(path); err == nil {
+						err = mknod(path, chrMode(mode), int(mkdev(devMajor, devMinor)))
+					}
+				}
+			case tar.TypeBlock:
+				err = mknod(path, blkMode(mode), int(mkdev(devMajor, devMinor)))
+				if err != nil && os.IsExist(err) {
+					if err = os.Remove(path); err == nil {
+						err = mknod(path, blkMode(mode), int(mkdev(devMajor, devMinor)))
+					}
+				}
+			case tar.TypeDir:
+				err = os.Mkdir(path, mode)
+				if err != nil && os.IsExist(err) {
+					err = nil
+				}
+				// make a note of the directory's times.  we
+				// might create items under it, which will
+				// cause the mtime to change after we correct
+				// it, so we'll need to correct it again later
+				directoriesAndTimes = append(directoriesAndTimes, directoryAndTimes{
+					directory: path,
+					atime:     hdr.AccessTime,
+					mtime:     hdr.ModTime,
+				})
+			case tar.TypeFifo:
+				fifoMode := uint32(hdr.Mode)
+				err = mkfifo(path, fifoMode)
+				if err != nil && os.IsExist(err) {
+					if err = os.Remove(path); err == nil {
+						err = mkfifo(path, fifoMode)
+					}
+				}
+			}
+			// check for errors
+			if err != nil {
+				return errors.Wrapf(err, "error creating %q", path)
+			}
+			if written != hdr.Size {
+				return errors.Errorf("error creating %q: incorrect length (%d != %d)", path, written, hdr.Size)
+			}
+			// restore xattrs
+			if !request.PutOptions.StripXattrs {
+				if err = Lsetxattrs(path, hdr.Xattrs); err != nil { // nolint:staticcheck
+					if !request.PutOptions.IgnoreXattrErrors {
+						return errors.Wrapf(err, "error setting extended attributes on %q", path)
+					}
+				}
+			}
+			// restore permissions, except for symlinks, since we don't have lchmod
+			if hdr.Typeflag != tar.TypeSymlink {
+				if err = os.Chmod(path, mode); err != nil {
+					return errors.Wrapf(err, "error setting permissions on %q to 0%o", path, mode)
+				}
+			}
+			// set ownership
+			if err = os.Lchown(path, hdr.Uid, hdr.Gid); err != nil {
```

Comment:
If this is supposed to be safe against concurrent access, create with 0700 or less; `Lchown; Chmod`.

Reply:
Changed it.  Thanks!